Fixes Sampling Markov Chain PR#791 #789

kris-singh opened the pull request:
This PR solves solves 2 Issues #791 Sampling problem and #789 Stationarity Test for Distribution. In Stationarity Test we check against steady state values obtained from the Eigen value and check the steady state distribution obtained from the samples that we generate these values have to be within a user defined parameter called tolerance to constitute a Passing of the stationarity test.

Commit:
Fixes Sampling Markov Chain

kris-singh commented:
i will also solve issue #789 in this request. As soon as we get a clear way of finding the stationarity condition is satisfied.

kris-singh commented:
@ankurankan can you review this and also for the stationarity condition i am thinking of implementing that as rabbe's test for convergence of the series

Commit:
Stationarity Test

kris-singh commented:
@ankurankan @yashu-seth @khalibartan  this is ready for review..

khalibartan commented on pgmpy/models/MarkovChain.py:
```diff
@@ -367,6 +369,48 @@ def generate_sample(self, start_state=None, size=1):
                 self.state[j] = State(var, next_st)
             yield self.state[:]
 
+    def check_stationarity(self, tolerance=0.2, sample=None):
```
@kris-singh @ankurankan IMO instead of raising a `ValueError` when markov chain is not stationary, it should return a Boolean value. Or is there any particular reason for this design?

ankurankan replied:
@kris-singh I think the method name should be `is_stationary` and it should return True or False.

khalibartan commented on pgmpy/models/MarkovChain.py:
```diff
@@ -367,6 +369,48 @@ def generate_sample(self, start_state=None, size=1):
+        Checks if the given markov model is stationary and checks the steady state
+        probablity values for the state are consistent.
+
+
```
@kris-singh Remove extra blank line.

Commit:
Reviews Updated

ReviewChanges Made

Change retunr type

Review Added

Commits squash

Version Error fixed

Version Error Fixed

kris-singh commented:
@khalibartan your comments are updated. @ankurankan can you review this as i think this is high priority it solves the sampling error #791.

yashu-seth commented on pgmpy/models/MarkovChain.py:
```diff
@@ -367,6 +369,49 @@ def generate_sample(self, start_state=None, size=1):
+        >>> diff_tm = {0: {0: 0.5, 1: 0.5}, 1: {0: 0.25, 1:0.75}}
+        >>> model.add_transition_model('diff', diff_tm)
+        >>> model.check_stationarity()
+        """
```
@kris-singh Add the output for ```model.check_stationarity```

yashu-seth commented on pgmpy/models/MarkovChain.py:
```diff
@@ -4,6 +4,8 @@
 import numpy as np
 from pandas import DataFrame
+from scipy.linalg import eig
+import math
```
@kris-singh PEP8 
http://stackoverflow.com/questions/22722976/import-order-coding-standard

khalibartan commented on pgmpy/models/MarkovChain.py:
```diff
@@ -367,6 +368,50 @@ def generate_sample(self, start_state=None, size=1):
+
+        Return Type:
+        ------------
+        None
```
@kris-singh Return type is Boolean not none.

khalibartan commented on pgmpy/models/MarkovChain.py:
```diff
@@ -367,6 +368,50 @@ def generate_sample(self, start_state=None, size=1):
+            stationary = np.array(U[:, np.where(np.abs(S - 1.) < 1e-8)[0][0]].flat)
+            stationary = (stationary / np.sum(stationary)).real
+            probabilites = []
+            window_size = 10000 if sample == None else len(sample) 
```
@kris-singh Why have you chosen 10000 as a default length?

kris-singh replied:
This the default length previously used i didn't change it

khalibartan replied:
@kris-singh 10000 is really a large number for burn-in period. For some complex chains it can be time consuming. Also we haven't mentioned anything about this in docstring. @ankurankan What do you suggest here ?

I was thinking about using a probabilistic approach for choosing this window size.

ankurankan replied:
@khalibartan I think that the burn in period really depends on the distribution and the start state. So, we can't have a general number for that. What probabilistic approach are you suggesting ?

khalibartan replied:
@ankurankan Yes i know that burn-in period depends upon the start state and transition probabilities that's why I was suggesting not to use some large value. This might not sound good but i was thinking of using a random value out of a set of most likely values (like choosing a random value out of (100, 1000, 10000) or something like that.

ankurankan replied:
@khalibartan Yeah I agree we should have some smaller value. Or maybe by default we should have 0 burn in period. But I don't get the point of choosing a random value.

kris-singh replied:
Should we let the user decide the burn in period. with param like burn_in_period this can be categorical like small medium long

khalibartan commented:
@kris-singh Can you change the sections which are using pandas.Dataframe objects to use recarray as default? We want to make pandas as optional dependency. For reference you can checkout my PR #719

ankurankan commented on pgmpy/models/MarkovChain.py:
```diff
@@ -367,6 +368,50 @@ def generate_sample(self, start_state=None, size=1):
 
+    def check_stationarity(self, tolerance=0.2, sample=None):
+        """
+        Checks if the given markov model is stationary and checks the steady state
```
@kris-singh Use `Markov Chain` instead of markov model.

ankurankan commented on pgmpy/models/MarkovChain.py:
```diff
@@ -367,6 +368,50 @@ def generate_sample(self, start_state=None, size=1):
+
+        Return Type:
+        ------------
+        Boolean
```
@kris-singh Explain when do you expect True and when do you expect False.

ankurankan commented on pgmpy/models/MarkovChain.py:
```diff
@@ -367,6 +368,50 @@ def generate_sample(self, start_state=None, size=1):
+        """
+        keys = self.transition_models.keys()
+        for k in keys:
+            transition_mat = DataFrame(self.transition_models[k]).as_matrix()
```
@kris-singh Do you really need to convert it to a `DataFrame` ? We are trying to have pandas as an optional dependency.

khalibartan replied:
@kris-singh I think there is really no need of Dataframe here.

ankurankan commented on pgmpy/models/MarkovChain.py:
```diff
@@ -367,6 +368,50 @@ def generate_sample(self, start_state=None, size=1):
+            S, U = eig(transition_mat.T)
+            stationary = np.array(U[:, np.where(np.abs(S - 1.) < 1e-8)[0][0]].flat)
+            stationary = (stationary / np.sum(stationary)).real
+            probabilites = []
```
@kris-singh Add blank lines between different sections of code.

kris-singh commented:
@ankurankan i didn't understand why the check failed.
